packages/context/src/binding.ts

```diff
@@ -227,19 +269,32 @@ export class Binding {
       }
     }
     if (this._getValue) {
-      const result = this._getValue(ctx);
+      const resolutionSession = ResolutionSession.enterBinding(this, session);
+      let result = RejectionError.catch(this._getValue(ctx, resolutionSession));
+      result = RejectionError.then<BoundValue, BoundValue>(result, val => {
+        resolutionSession.exit();
```

Comment:
Your code here does not call `resolutionSession.exit` on error. I find it surprising, I would expect that resolution session should be finished (exited) on errors too. If it's not the case and the resolution session does not need to be touched on errors, then I am proposing to use a different name for `exit`, for example `bindingResolved` (as a counterpart to `enterBinding`).

In my experience, an explicit `enterBinding` + `exit` API pushes too much responsibility for correct error handling to the API callers. As a result, it often yields a code that misbehaves in case of errors not anticipated by the developer. A better solution is to write a wrapper function that 1) calls "enterBinding" 2) invokes passed callback to perform whatever should be performed as part of the session 3) calls "exit" when the callback is done, ensuring "exit" is called on errors too

```ts
class ResolutionSession {
  async execute(cb: () => Promise) {
    this.enterBinding();
    try {
      return await cb();
    } finally {
      this.exit();
    }
  }
}
```

Reply:
1. The RejectionError.catch() already converts rejected promises to normal ones. As a result, `resolutionSession.exit()` will cover both cases.

2. I don't think `finally` has been supported by TypeScript.